### PR TITLE
usb/acm: Blocking mode for USBUS CDC ACM STDIO

### PR DESCRIPTION
**Pull request by ant9000**
### Contribution description

I've modified USBUS CDC ACM STDIO to support blocking on pending characters from USB. The behaviour is controlled by CONFIG_USBUS_CDC_ACM_STDIO_BLOCKING variable, and can be configured via Kconfig.

### Testing procedure

I have tested it against my own [PoC](https://github.com/ant9000/test_usb_uart). With the default RIOT config, characters from USB are lost as soon as the STDIO buffer (defined by CONFIG_USBUS_CDC_ACM_STDIO_BUF_SIZE) is overrun.
If I enable the new blocking mode, the sender application slows down and adapts to the receiver speed.

### Issues/PRs references

This PR implements feature #16500 .

### CODE REVIEW

File: sys/usb/usbus/cdc/acm/cdc_acm_stdio.c [hunk 1]
```diff
@@ -75,8 +93,19 @@ static void _cdc_acm_rx_pipe(usbus_cdcacm_device_t *cdcacm,
...
     (void)cdcacm;
+#if CONFIG_USBUS_CDC_ACM_STDIO_BLOCKING
+    thread = thread_getpid();
+#endif
+
     for (size_t i = 0; i < len; i++) {
+#if CONFIG_USBUS_CDC_ACM_STDIO_BLOCKING
+        while (isrpipe_write_one(&_cdc_stdio_isrpipe, data[i]) < 0) {
+            xtimer_set(&timeout, XTIMER_BACKOFF+1); // shortest possible non-spinning wait
+            thread_sleep();
```
Comment (benpicco) [inline sys/usb/usbus/cdc/acm/cdc_acm_stdio.c:104]: how is this different from

```C
xtimer_sleep(XTIMER_BACKOFF + 1);
```

Comment (ant9000) [reply sys/usb/usbus/cdc/acm/cdc_acm_stdio.c:104]: Probably at all, if you say so :-)
I've tried xtimer_usleep but it was getting stuck, so I started digging further and came up with the awkward solution you see. I just did not realize it was only me waiting for too short a time, likely - I'll test againg.
Thanks for the tip!

Comment (ant9000) [reply sys/usb/usbus/cdc/acm/cdc_acm_stdio.c]: With your suggestion I've come up with a **much** cleaner solution. Thank you so much!

**Comment by benpicco** [comment]
Now this makes me wonder: why should this be optional?

**Comment by ant9000** [comment]
I'd say there are two possible problems. First, it behaves differently from what is now the default - so it might break something I'm not aware of. And second, my patch does not implement any way out: so if no one consumes the arriving bytes, the board will get stuck forever.

**Comment by bergzand** [comment]
> if no one consumes the arriving bytes, the board will get stuck forever.

This is why the original implementation is non-blocking. It is not always clear for a device if the USB port is disconnected from the host, latency is unpredictable, and things can break unpredictably if you add a blocking write to the USB thread code.

**Comment by ant9000** [comment]
Yes, and this is probably the best choice for the default value

> > if no one consumes the arriving bytes, the board will get stuck forever.
> 
> This is why the original implementation is non-blocking. It is not always clear for a device if the USB port is disconnected from the host, latency is unpredictable, and things can break unpredictably if you add a blocking write to the USB thread code.

I'd say the best choice for a default value is non-blocking, yes. That said, there are use cases where a blocking mode could be valuable - my PR aims at introducing that, in case one needs it (I do for a project, that's why I wrote it).

**Comment by benpicco** [comment]
Is there no way to detect if a USB device is connected to a host?

**Comment by ant9000** [comment]
> > if no one consumes the arriving bytes, the board will get stuck forever.
> 
> This is why the original implementation is non-blocking. It is not always clear for a device if the USB port is disconnected from the host, latency is unpredictable, and things can break unpredictably if you add a blocking write to the USB thread code.

On second thought, I missed to mention one point: this PR addresses _reads_, not writes. I want to avoid overrunning the  buffer for stdio_read.

**Comment by ant9000** [comment]
> Is there no way to detect if a USB device is connected to a host?

AFAIK, this is what `USBUS_CDC_ACM_LINE_STATE_DISCONNECTED` should be for - see [here](https://github.com/RIOT-OS/RIOT/blob/137dd14911b27eeb4a729c29efb0c0472ec3ee27/sys/usb/usbus/cdc/acm/cdc_acm.c#L153) for instance. I really don't how well this works in practice on the various USB implementations supported by RIOT. I can vouch that it works as intended on SAML21 and SAMR34 CPUs, on which I've tested it repeatedly.

**Comment by ant9000** [comment]
Sorry to bug you: did anybody have the chance of testing my PR? Any problems/comments/suggestions?

Thanks a lot,

Antonio

**Comment by kfessel** [comment]
@ant9000 isn't there another mutex you could wait on (eg the read of the buffer) instead of a "just not spinning xtimer" 
I think this will eat up much(most) of your cpu in thread switching.

**Comment by ant9000** [comment]
> @ant9000 isn't there another mutex you could wait on (eg the read of the buffer) instead of a "just not spinning xtimer"
> I think this will eat up much(most) of your cpu in thread switching.

I fear what you suggest is not possible. My idea is that I need to wait until the available data can be **written** to the buffer. If I lock the CPU on this task, no other thread can consume what is already there and the board deadlocks - so I'd say that thread switching is absolutely necessary for things to work as intended.

But please, do correct me if my reasoning is wrong!

**Comment by kfessel** [comment]
you would need a new mutex (not the one inside the isr_pipe) (instead of a new timer)
lock it if you are not able to write to the buffer and unlock it if when data is read from the buffer (you can unlock it every time as the unlock performs  a check if it was locked) and you just lock it if you cloud not write (as the test write to isr_pipe is safe).

atm i am not sure where the `_cdc_acm_rx_pipe` is run  i would assume it run in the usb event  thread?
therefor all other threads are free to run and the usb event  thread is blocked (like it is in this with the timer)

**Comment by ant9000** [comment]
> you would need a new mutex (not the one inside the isr_pipe) (instead of a new timer)
> lock it if you are not able to write to the buffer and unlock it if when data is read from the buffer (you can unlock it every time as the unlock performs a check if it was locked) and you just lock it if you cloud not write (as the test write to isr_pipe is safe).
> 
> atm i am not sure where the `_cdc_acm_rx_pipe` is run i would assume it run in the usb event thread?
> therefor all other threads are free to run and the usb event thread is blocked (like it is in this with the timer)

Your suggestion makes a lot of sense. I've rewritten the code, and it indeed works faster.

Thank you very much!

**Comment by kfessel** [comment]
I think the Patch does what it intends to do. :heavy_check_mark:

I am not sure about its behavior in relation to client usb-stack and the host if no one reads the stdio (or very slow); I think this will lock the USB eventloop:
    Is this ok for the rest of the USB system? (Hubs and hosts?)
    Will connections be dropped? (just ACM? / the Device? / the Hub?)

I think this need some documentation especially in these points (Maybe it needs a big warning).

@bergzand seems to the person that might be able to answer and review these points.

other than that i like what this patch does (no one likes dropped bytes) and i like its now reasonable impact.

**Comment by kfessel** [comment]
@ant9000: I did not yet find the time to check what happens in the no reader / slow reader case with 
Did you do such checks?

Maybe you are able to answer:
Is this ok for the rest of the USB system? (Hubs and hosts?)
Will connections be dropped? (just ACM? / the Device? / the Hub?)


as for some riot formalities:
- please add "sys/usb/acm:" or "usb/acm:" at the front of your commit headline (see head of this PR)
- please avoid force pushes to your PR: Head (unless required (renaming/squashing/rebasing))

- please remove the (see ... ) from your commit message as it clutters your issue, you may name the issue for documentation reasons in the Commit-body

**Pull request closed without merging**